src/sources/kafka.rs

```diff
@@ -370,105 +411,559 @@ impl SourceConfig for KafkaSourceConfig {
+                        } else {
+                            state.keep_draining(drain_deadline)
+                        }
+                    },
+                    ConsumerState::Consuming(state) => {
+                        // If we are here, it is likely because the consumer
+                        // tasks are set up to exit upon reaching the end of the
+                        // partition.
+                        if !exit_eof {
+                            debug!("Partition consumer task finished, while not in draining mode.");
+                        }
+                        state.keep_consuming(drain_deadline)
+                    },
+                };
+
+                match status {
+                    // PartitionConsumerStatus differentiates between a task that exited after
+                    // being signaled to end, and one that reached the end of its partition and
+                    // was configured to exit. After the last such task ends, we signal the kafka
+                    // driver task to shut down the main consumer too. Note this is only used in tests.
+                    PartitionConsumerStatus::PartitionEOF if exit_eof && partition_consumers.is_empty() => {
+                        debug!("All partitions have exited or reached EOF.");
+                        let _ = eof.take().map(|e| e.send(()));
+                    },
+                    _ => {}
                 }
             },
-            message = stream.next() => match message {
-                None => break,  // WHY?
-                Some(Err(error)) => emit!(KafkaReadError { error }),
-                Some(Ok(msg)) => {
-                    emit!(KafkaBytesReceived {
-                        byte_size: msg.payload_len(),
-                        protocol: "tcp",
-                        topic: msg.topic(),
-                        partition: msg.partition(),
-                    });
+            Some(callback) = callbacks.recv() => match callback {
+                KafkaCallback::PartitionsAssigned(mut assigned_partitions, done) => match consumer_state {
+                    ConsumerState::Complete(_) => unreachable!("Partition assignment received after completion."),
+                    ConsumerState::Draining(_) => error!("Partition assignment received while draining revoked partitions, invalid assignment?"),
+                    ConsumerState::Consuming(ref consumer_state) => {
+                        let acks = consumer.context().acknowledgements;
+                        for tp in assigned_partitions.drain(0..) {
+                            let topic = tp.0.as_str();
+                            let partition = tp.1;
+                            // It _should_ be impossible for this expect() to panic, since we receive the topic/partition pair from the rebalance callback
+                            let pq = consumer.split_partition_queue(topic, partition).expect("Failed to get partition queue: invalid topic or partition.");
```

Comment:
Would it be safer to use `if let Some(pq) = consumer.split_partition_queue(topic, partition) { … }` to avoid the panic, since we don't control that part of the code?

Reply:
Yeah, we can emit a warning log in case this ever returns `None`